File: majd2mm.pl
#!/usr/bin/perl -w 

# Convert majordomo email commands to mailman email commands
# Jomo, 6/4/2013

my $email_body="";
my $cmd;
my $domainName = "linus.netazoic.com";
my $line;
my $from;

my $cmdSub = "http://[domain.name]/mailman/admin/[listname]/members/add?subscribees=[user_email_address]&adminpw=[admin_password]&send_welcome_msg_to_this_batch=0&send_notifications_to_list_owner=0";
my $cmdUnsub = "http://[domain.name]/mailman/options/[listname]/[user_email_address]?unsub=1&unsubconfirm=1&password=[admin_password]";

my $cmdMM;
my $logMsg;
my $flgUnsub = 0;


#************* SET THESE VALUES **************
my $MM_BIN="/usr/lib/mailman/bin";
my $MEMBERS_FILE="./members.txt";
my $LIST="ccb_genrl";
my $LOG_FILE="/var/log/majd2mm.log";
#********************************************

if(! -d $MM_BIN){ exit("The mailman bin directory not located. Please edit settings.");}


sub processCmd($){
	my($cmd1) = @_;
	my $ret;

	chomp($cmd1);


	if($cmd1 =~ /^approve.*/){
		$cmd1 =~ /approve\s+([\w|\.]+)\s+([\w]+)\s+([\w|_|\.|\*]+)\s+([\w|_|@|\.]+)/;
		$passwd = $1;
		$cmdVerb = $2;
		$list = $3;
		$email = $4;
	}
	elsif($cmd1 =~ /^unsubscribe\s+([\w|_|\*]+)\s+(.*)/){
		#$cmd1 =~ /unsubscribe\s+([\w|_|\*]+)\s+([\w|@|_|\.|\*]+)/;
		$cmdVerb = "unsubscribe";
		$list = $1;
		$email = $2;
	}
	else{
		$cmdVerb = "ERROR";
		$logMsg = "problem processing: $cmd1";
		return;
	}



	# Basic authentication is killing the wget idea
	#if($cmdVerb eq "subscribe"){ 
	#	$mmEmail = "join-$list";
	#	$cmdMM = $cmdSub;
	#}
	#if($cmdVerb=~/unsub/i){ 
	#	$mmEmail = 
	#	$cmdMM = $cmdUnsub;
	#}
	#$cmdMM =~s/\[domain.name\]/$domainName/;
	#$cmdMM =~s/\[listname\]/$list/;
	#$cmdMM =~s/\[user_email_address\]/$email/;
	#$cmdMM =~s/\[admin_password\]/$passwd/;
	#system("wget $cmdMM");



	# Email options aren't all that great
	#$mmEmail = $cmdVerb . "_" . $list;
	#$msg = "subscribe address=$email";	
	#$from = $majordomoEmail;
	#send_email($mmEmail,$from,$subject,$msg);

	# Just run commands directly

	&writeMembers($email);
	if($cmdVerb eq "subscribe"){
	
		# just run the add_members command
		$cmd = "$MM_BIN/add_members -r $MEMBERS_FILE $LIST";
		$cmd .=" -w y -a n";
		$cmd .= " $list";
		$ret = `$cmd`;
		if(! $?){$logMsg = $list . ": " . $ret}
		else{	$logMsg = "Error while trying to subscribe $email to $list\n";
			$logMsg .= "$ret\n";
			}
	}
	elsif($cmdVerb eq "unsubscribe"){

		$email=~s/\*//g;  # remove wildcards in the email address
		$cmd = "$MM_BIN/remove_members ";
		if($list eq "*"){ $cmd .= " --fromall";}
		else{ $cmd .= " $list ";}
		$cmd .= " $email";
		$ret = `$cmd`;
		if(! $?){
			if($ret){$logMsg = $ret;}
			#else{$logMsg = "Ran unsubscribe command with no results: $cmd";}
			else{$logMsg = "Unsubscribed $email from $list";}
		}
		else{	$logMsg = "Error while trying to unsubscribe $email from $list\n";
			$logMsg .= "$ret\n";
		}
	   
	}
	&writeLog($LOG_FILE, $logMsg);

}
	
	

while (<STDIN>){

	if(1){print($_);}
	$email_body .= $_;
	$line = $_;
	if(! $line){ next;}
	chomp($line);
	if(! $line){ next;}
	if ($line=~/^\s*From:.*[\s:\<]\s*([\w\.]+@[\w\.]+)/){
		$from = $1;next;
	}
	if($line=~/^\s+?approve.*/){
		$cmd = $line;
		processCmd($cmd);
		next;
	}
	if($line=~/^\s+?unsubscribe.*/){
		$cmd = $line;
		processCmd($cmd);
		next;
	}
	if($line=~/Subject:.*(unsub|remove)/i){
		$flgUnsub = 1;

		#Have to process this command outside the main loop as we can't be sure
		#the subject line will follow the From line

		#$cmd = "approve ccb.passwd unsubscribe * $from";
		#processCmd($cmd);
		next;
	}
	if(/^end$/){ 
		## How do we do a last?
		while(<STDIN>){
			if(0){print($_);}
		}
		next;
	}
	next;
	
}	

if($flgUnsub && $from){
# Special purpose -- process a user submitted remove or unsub command
		$cmd = "approve ccb.passwd unsubscribe * $from";
		processCmd($cmd);
		$flgUnsub = 0;
}



sub writeLog{
# write anything out to a log file
        ($logfile,$logtext) = @_;
	chomp($logtext);
	if($logtext!~/.*\n$/){
		$logtext .= "\n";
	}
        open (FH,">>$logfile");
        print FH $logtext;
        close FH;
        return;
}


sub writeMembers{
	my $memfile = $MEMBERS_FILE;
	my ($members) = @_;
	unlink($memfile);	# zero it out
	open(FH, ">>$memfile");
	print(FH $members);
	close FH;
	return;
}

=begin GHOSTCODE
=end GHOSTCODE

File: majd2mm.py
#!/usr/bin/python
'''
Created on Jul 11, 2022

@author: John Moore,  jmoore@netazoic.com

majd2mm.majd2mm -- Add majordomo style email processing to the mailman system

majd2mm.majd2mm is a conversion of the majd2mm.pl perl script, 
             the purpose of which is to convert majordomo email commands to mailman email commands
             Primarily used for unsubscribe
'''

import os
import re
import subprocess
import time
import sys
import datetime
import pwd

#import pdb
#pdb.set_trace()

email_body = "";
cmd = "";
domainName = "mm2.freeburmarangers.org";
line = "";
fromAddr = "";

cmdMM = "";
logMsg = "";
flgUnsub = 0;
flgDebug = 0;


#************* SET THESE VALUES **************
MM_BIN = "/usr/lib/mailman/bin";
MEMBERS_FILE = "./members.txt";
LIST = "ccb_genrl";
LOG_FILE = "/var/log/majd2mm.log";
TEST_FILE = "./test/test.forward.txt"
user_name = "fbr.mailadmin"
#********************************************
binFileExists = os.path.isdir(MM_BIN)
if(not binFileExists):
     if(not flgDebug): 
         exit("The mailman bin directory not located. Please edit settings.");
         
pw_record = pwd.getpwnam(user_name)
user_name      = pw_record.pw_name
user_home_dir  = pw_record.pw_dir
user_uid       = pw_record.pw_uid
user_gid       = pw_record.pw_gid

def writeLog(logfile, logtext):
# write anything out to a log file
    logTime = datetime.datetime.now().strftime("%I:%M%p on %B %d, %Y")
    logtext = logtext.rstrip()
    # makesure logtext always ends with a newline
    logtext += "\n";  
    logtext = logTime + " " + logtext;
    if(flgDebug): print(logtext);
    f = open(logfile,'a');
    f.write(logtext);
    f.close();


def writeMembers(members):
    memfile = MEMBERS_FILE;
    f = open(memfile)
    # zero it out
    f.delete();    
    f.write(members);
    f.close();


def processCmd(cmd):
    lists = []

    if(re.search("^approve.*", cmd)):
        regex = "approve\s+([\w|\.]+)\s+([\w]+)\s+([\w|_|\.|\*]+)\s+([\w|_|@|\.]+)"
        m = re.search(regex, cmd);
        passwd = m.group(1);
        cmdVerb = m.group(2);
        listName = m.group(3);
        email = m.group(4);

    elif(re.search("^unsubscribe.*", cmd)):
        regex = "^unsubscribe\s+([\w|_|\*|\|]+)\s+(.*)"
        m = re.search(regex, cmd);
        cmdVerb = "unsubscribe";
        listName = m.group(1);
        email = m.group(2);

    else:
        cmdVerb = "ERROR";
        logMsg = "problem processing: " + cmd;
        return;

    # Just run commands directly
    
    if(cmdVerb == "subscribe"):
        # just run the add_members command
        # cmd = " %s/add_members -r %s " %(MM_BIN,MEMBERS_FILE);
        cmd = " %s/sub " % (MM_BIN);
        cmd += " -w y -a n ";
        cmd += listName + " " + email
	try:
            ret = subprocess.check_output([cmd,'-w y', '-a n', listname, email])
            logMsg = listName + ": " + ret
	    next
        except subprocess.CalledProcessError:
            logMsg = "Error while trying to subscribe %s to %s\n" % (email, listName);
            logMsg += ret;
            
    elif(cmdVerb == "unsubscribe"):
        email = email.replace("\*", "");  # remove wildcards in the email address
        cmd = "%s/remove_members " % MM_BIN;
        if(listName == "*"):
            cmd += " --fromall";
            lists = [listName]
        else:
            if(re.match(".*\|", listName)):
                lists = listName.split("|")
            else:
                lists = [listName]
        for ln in lists:
	    result = ""
            cmd1 = cmd;
            cmd1 += ln + " "
            cmd1 += email
	    try:
		result = subprocess.check_output(cmd1.split(" "), preexec_fn=demote(user_uid, user_gid))
                logMsg = "Unsubscribed %s from %s" % (email, ln)
		next
	    except subprocess.CalledProcessError:
                logMsg = "Error while trying to unsubscribe %s from %s\n" % (email, ln);
                logMsg += "%s \n" % result 
        
            writeLog(LOG_FILE, logMsg);
    
    ### End of processCmd
    
def demote(user_uid, user_gid):
   def result():
     os.setgid(user_gid)
     os.setuid(user_uid)
   return result
    
## Process the FILE (email)


if(flgDebug):
    inputFile = TEST_FILE
else:
    inputFile = sys.stdin
    
#for line in open(inputFile):
for line in sys.stdin:
    if(True): 
        print(line)
    email_body += line
    line = line.rstrip();
    m = re.search("^\s*From:.*[\s:\<]\s*([\w\.]+@[\w\.]+)", line)
    if(m):
        fromAddr = m.group(1)
        next
    # check for a command
    m = re.search("^\s*?(approve|unsubscribe).*", line)
    if(m):
        cmd = line
        processCmd(cmd)
        continue

    # check for an unsubscribe in email
    m = re.search("^\s*?Subject:.*(unsub|remove)", line, re.IGNORECASE)
    #m = re.search("^Subject:.*", line)
    if(m):
        flgUnsub = 1;
        # Have to process this command outside the main loop as we can't be sure
        # the subject line will follow the From line
        # $cmd = "approve ccb.passwd unsubscribe * $from";
        # processCmd($cmd);
        next;

    m = re.search("^\s*remove\s*", line, re.IGNORECASE)
    if(m):
        flgUnsub = 1;
        next;

    # check for an END_PROCESSING command
    m = re.search("^end$", line)
    if(m):
        break
    
    next;

if(flgUnsub and not fromAddr.isspace()):
# Special purpose -- process a user submitted remove or unsub command
        cmd = "unsubscribe ccb_genrl|ccb_specl|ccb_prayer %s" % fromAddr;
        processCmd(cmd);
        flgUnsub = 0;

